File: app/src/main/java/com/icerrate/popularmovies/view/movies/catalog/MoviesCatalogFragment.java
package com.icerrate.popularmovies.view.movies.catalog;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.ImageView;
import android.widget.TextView;

import com.icerrate.popularmovies.R;
import com.icerrate.popularmovies.data.model.Movie;
import com.icerrate.popularmovies.data.model.PaginatedResponse;
import com.icerrate.popularmovies.utils.InjectionUtils;
import com.icerrate.popularmovies.utils.ViewUtils;
import com.icerrate.popularmovies.view.common.BaseFragment;
import com.icerrate.popularmovies.view.common.EndlessRecyclerOnScrollListener;
import com.icerrate.popularmovies.view.movies.detail.MovieDetailActivity;
import com.icerrate.popularmovies.view.movies.catalog.MoviesCatalogPresenter.SortType;
import com.icerrate.popularmovies.view.movies.search.SearchMoviesActivity;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

import static com.icerrate.popularmovies.view.movies.detail.MovieDetailFragment.KEY_MOVIE;

/**
 * @author Ivan Cerrate.
 */

public class MoviesCatalogFragment extends BaseFragment implements MoviesCatalogContract.View,
        MoviesCatalogAdapter.OnItemClickListener {

    public final static int RC_MOVIE_DETAIL = 25;

    public final static String KEY_SORT_TYPE = "SORT_TYPE_KEY";

    public final static String KEY_PAGINATED_MOVIES = "PAGINATED_MOVIES_KEY";

    @BindView(R.id.movies)
    public RecyclerView moviesRecyclerView;

    @BindView(R.id.movies_no_data)
    public TextView noDataTextView;

    @BindView(R.id.footer_progress)
    public ViewStub footerProgressBar;

    private MoviesCatalogAdapter adapter;

    private EndlessRecyclerOnScrollListener endlessRecyclerOnScrollListener;

    private MoviesCatalogPresenter presenter;

    private ImageView selectedMoviePoster;

    public static MoviesCatalogFragment newInstance() {
        return new MoviesCatalogFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        presenter = new MoviesCatalogPresenter(this,
                InjectionUtils.movieRepository(getContext()));
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        View view = inflater.inflate(R.layout.fragment_movies_catalog, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setupView();
        if (savedInstanceState != null) {
            restoreInstanceState(savedInstanceState);
            presenter.loadMovies();
        } else {
            presenter.loadMoviesBySortType(SortType.MOST_POPULAR);
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_movies_catalog, menu);
        MoviesCatalogPresenter.SortType currentSortType = presenter.getSortType();
        if (currentSortType.equals(MoviesCatalogPresenter.SortType.MOST_POPULAR)) {
            MenuItem item = menu.findItem(R.id.popular);
            item.setChecked(true);
        } else if (currentSortType.equals(MoviesCatalogPresenter.SortType.HIGHEST_RATED)) {
            MenuItem item = menu.findItem(R.id.top_rated);
            item.setChecked(true);
        } else if (currentSortType.equals(MoviesCatalogPresenter.SortType.FAVORITE)) {
            MenuItem item = menu.findItem(R.id.favorite);
            item.setChecked(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        SortType selectedSortType;
        switch (item.getItemId()) {
            case R.id.popular:
                selectedSortType = SortType.MOST_POPULAR;
                break;
            case R.id.top_rated:
                selectedSortType = SortType.HIGHEST_RATED;
                break;
            case R.id.favorite:
                selectedSortType = SortType.FAVORITE;
                break;
            case R.id.action_search:
                startActivity(SearchMoviesActivity.makeIntent(getContext()));
                getActivity().overridePendingTransition(R.anim.fade_in, R.anim.none);
                return true;
            default:
                return false;
        }
        presenter.refreshMoviesBySortType(selectedSortType);
        item.setChecked(true);
        return true;
    }

    @Override
    protected void saveInstanceState(Bundle outState) {
        outState.putSerializable(KEY_SORT_TYPE, presenter.getSortType());
        outState.putParcelable(KEY_PAGINATED_MOVIES, presenter.getMoviesPaginatedResponse());
    }

    @Override
    protected void restoreInstanceState(Bundle savedState) {
        MoviesCatalogPresenter.SortType sortType = (MoviesCatalogPresenter.SortType) savedState.getSerializable(KEY_SORT_TYPE);
        PaginatedResponse<Movie> moviesPaginatedResponse = savedState.getParcelable(KEY_PAGINATED_MOVIES);
        presenter.loadPresenterState(sortType, moviesPaginatedResponse);
    }

    private void setupView() {
        //Movies
        int columns = ViewUtils.getGridColumns(getActivity());
        GridLayoutManager gridLayoutManager = new GridLayoutManager(getContext(), columns);
        endlessRecyclerOnScrollListener = new EndlessRecyclerOnScrollListener(gridLayoutManager) {
            @Override
            public void onLoadMore() {
                presenter.loadNextMoviesPage();
            }
        };
        adapter = new MoviesCatalogAdapter(this, columns);
        moviesRecyclerView.setAdapter(adapter);
        moviesRecyclerView.setLayoutManager(gridLayoutManager);
        moviesRecyclerView.addOnScrollListener(endlessRecyclerOnScrollListener);
        //refresh
        if (refreshLayout != null) {
            refreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
                @Override
                public void onRefresh() {
                    presenter.refreshMovies();
                }
            });
            refreshLayout.setColorSchemeColors(getResources().getColor(R.color.colorAccent));
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == RC_MOVIE_DETAIL && resultCode == Activity.RESULT_OK) {
            presenter.onBackFromDetail();
        }
    }

    @Override
    public void resetMovies() {
        adapter.resetItems();
        adapter.setLoading(false);
        endlessRecyclerOnScrollListener.setLoading(false);
    }

    @Override
    public void showMovies(ArrayList<Movie> movies) {
        adapter.addItems(movies);
        endlessRecyclerOnScrollListener.setLoading(false);
    }

    @Override
    public void showNoDataView(boolean show) {
        noDataTextView.setVisibility(show ? View.VISIBLE : View.GONE);
        noDataTextView.setText(getString(R.string.movies_no_data));
    }

    @Override
    public void showFooterProgress(boolean show) {
        adapter.setLoading(show);
        footerProgressBar.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @Override
    public void goToMovieDetail(Movie movie) {
        ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(getActivity(),
                selectedMoviePoster, getString(R.string.movie_poster_transition));
        startActivityForResult(MovieDetailActivity.makeIntent(getActivity())
                .putExtra(KEY_MOVIE, movie), RC_MOVIE_DETAIL, options.toBundle());
        getActivity().overridePendingTransition(R.anim.fade_in, R.anim.none);
    }

    @Override
    public void onItemClick(View view) {
        Movie movie = (Movie) view.getTag();
        if (movie != null) {
            selectedMoviePoster = view.findViewById(R.id.poster);
            presenter.onMovieItemClick(movie);
        }
    }
}

File: app/src/main/java/com/icerrate/popularmovies/view/movies/catalog/MoviesCatalogFragment.kt
package com.icerrate.popularmovies.view.movies.catalog

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.widget.ImageView
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.app.ActivityOptionsCompat
import androidx.core.view.MenuProvider
import androidx.lifecycle.Lifecycle
import androidx.recyclerview.widget.GridLayoutManager
import com.icerrate.popularmovies.databinding.FragmentMoviesCatalogBinding
import com.icerrate.popularmovies.R
import com.icerrate.popularmovies.data.model.Movie
import com.icerrate.popularmovies.data.model.PaginatedResponse
import com.icerrate.popularmovies.utils.ViewUtils
import com.icerrate.popularmovies.data.source.MovieRepository
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import com.icerrate.popularmovies.view.common.BaseFragment
import com.icerrate.popularmovies.view.common.EndlessRecyclerOnScrollListener
import com.icerrate.popularmovies.view.movies.catalog.MoviesCatalogPresenter.SortType
import com.icerrate.popularmovies.view.movies.detail.MovieDetailActivity
import com.icerrate.popularmovies.view.movies.detail.MovieDetailFragment.Companion.KEY_MOVIE
import com.icerrate.popularmovies.view.movies.search.SearchMoviesActivity

/**
 * @author Ivan Cerrate.
 */
@AndroidEntryPoint
class MoviesCatalogFragment : BaseFragment<FragmentMoviesCatalogBinding>(
    FragmentMoviesCatalogBinding::inflate
), MoviesCatalogContract.View,
    MoviesCatalogAdapter.OnItemClickListener, MenuProvider {

    private lateinit var adapter: MoviesCatalogAdapter
    private lateinit var endlessRecyclerOnScrollListener: EndlessRecyclerOnScrollListener
    private lateinit var presenter: MoviesCatalogPresenter
    private var selectedMoviePoster: ImageView? = null
    
    private lateinit var movieDetailLauncher: ActivityResultLauncher<Intent>
    
    @Inject
    lateinit var movieRepository: MovieRepository

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        presenter = MoviesCatalogPresenter(
            this,
            movieRepository
        )
        
        // Initialize the ActivityResultLauncher
        movieDetailLauncher = registerForActivityResult(
            ActivityResultContracts.StartActivityForResult()
        ) { result: ActivityResult ->
            if (result.resultCode == Activity.RESULT_OK) {
                presenter.onBackFromDetail()
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        requireActivity().addMenuProvider(this, viewLifecycleOwner, Lifecycle.State.RESUMED)
        setupView()
        if (savedInstanceState != null) {
            restoreInstanceState(savedInstanceState)
            presenter.loadMovies()
        } else {
            presenter.loadMoviesBySortType(SortType.MOST_POPULAR)
        }
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        menuInflater.inflate(R.menu.menu_movies_catalog, menu)
        val currentSortType = presenter.getSortType()
        when (currentSortType) {
            SortType.MOST_POPULAR -> menu.findItem(R.id.popular).isChecked = true
            SortType.HIGHEST_RATED -> menu.findItem(R.id.top_rated).isChecked = true
            SortType.FAVORITE -> menu.findItem(R.id.favorite).isChecked = true
        }
    }

    override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
        val selectedSortType = when (menuItem.itemId) {
            R.id.popular -> SortType.MOST_POPULAR
            R.id.top_rated -> SortType.HIGHEST_RATED
            R.id.favorite -> SortType.FAVORITE
            R.id.action_search -> {
                startActivity(SearchMoviesActivity.makeIntent(requireContext()))
                activity?.overridePendingTransition(R.anim.fade_in, R.anim.none)
                return true
            }
            else -> return false
        }
        presenter.refreshMoviesBySortType(selectedSortType)
        menuItem.isChecked = true
        return true
    }

    override fun saveInstanceState(outState: Bundle) {
        outState.putSerializable(KEY_SORT_TYPE, presenter.getSortType())
        outState.putParcelable(KEY_PAGINATED_MOVIES, presenter.getMoviesPaginatedResponse())
    }

    override fun restoreInstanceState(savedState: Bundle) {
        val sortType = savedState.getSerializable(KEY_SORT_TYPE) as SortType
        val moviesPaginatedResponse = savedState.getParcelable<PaginatedResponse<Movie>>(KEY_PAGINATED_MOVIES)
        presenter.loadPresenterState(sortType, moviesPaginatedResponse)
    }

    private fun setupView() {
        // Movies
        val columns = ViewUtils.getGridColumns(requireActivity())
        val gridLayoutManager = GridLayoutManager(context, columns)
        endlessRecyclerOnScrollListener = object : EndlessRecyclerOnScrollListener(gridLayoutManager) {
            override fun onLoadMore() {
                presenter.loadNextMoviesPage()
            }
        }
        adapter = MoviesCatalogAdapter(onItemClickListener = this, columns = columns)
        binding.movies.adapter = adapter
        binding.movies.layoutManager = gridLayoutManager
        binding.movies.addOnScrollListener(endlessRecyclerOnScrollListener)
        
        // Refresh
        binding.refresh.apply {
            setOnRefreshListener { presenter.refreshMovies() }
            setColorSchemeColors(resources.getColor(R.color.colorAccent))
        }
    }


    override fun resetMovies() {
        adapter.resetItems()
        adapter.setLoading(false)
        endlessRecyclerOnScrollListener.setLoading(false)
    }

    override fun showMovies(movies: ArrayList<Movie>) {
        adapter.addItems(movies)
        endlessRecyclerOnScrollListener.setLoading(false)
    }

    override fun showNoDataView(show: Boolean) {
        binding.moviesNoData.visibility = if (show) View.VISIBLE else View.GONE
        binding.moviesNoData.text = getString(R.string.movies_no_data)
    }

    override fun showFooterProgress(show: Boolean) {
        adapter.setLoading(show)
        binding.footerProgress.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun goToMovieDetail(movie: Movie) {
        selectedMoviePoster?.let { poster ->
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(
                requireActivity(),
                poster,
                getString(R.string.movie_poster_transition)
            )
            val intent = MovieDetailActivity.makeIntent(requireActivity()).putExtra(KEY_MOVIE, movie)
            movieDetailLauncher.launch(intent, options)
        }
        activity?.overridePendingTransition(R.anim.fade_in, R.anim.none)
    }

    override fun onItemClick(view: View) {
        val movie = view.tag as? Movie
        movie?.let {
            selectedMoviePoster = view.findViewById(R.id.poster)
            presenter.onMovieItemClick(it)
        }
    }

    override fun showProgressBar(show: Boolean) {
        binding.progress.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun showRefreshLayout(show: Boolean) {
        binding.refresh.isRefreshing = show
    }

    companion object {
        const val KEY_SORT_TYPE = "SORT_TYPE_KEY"
        const val KEY_PAGINATED_MOVIES = "PAGINATED_MOVIES_KEY"

        fun newInstance(): MoviesCatalogFragment = MoviesCatalogFragment()
    }
}